autoload/maktaba/syscall.vim

```diff
@@ -89,6 +121,66 @@ function! maktaba#syscall#DoCall() abort dict
   return l:return_data
 endfunction
 
+
+""
+" @public
+" Returns whether the current vim session supports asynchronous calls.
+function! maktaba#syscall#IsAsyncAvailable()
+  return !s:async_disabled && !empty(v:servername) && has('clientserver')
+endfunction
+
+
+""
+" @private
+" @dict Syscall
+" Calls |system()| asynchronously, and invokes a @function(this.callback) once
+" the command completes, passing in stdout, stderr and exit code to it.
+" The specific implementation for @function(#CallAsync).
+function! maktaba#syscall#DoCallAsync() abort dict
+  if !maktaba#syscall#IsAsyncAvailable()
+    if self.allow_sync_fallback
+      let l:return_data = self.Call()
+      if has_key(l:return_data, 'stderr')
+        let l:stderr = l:return_data.stderr
+      else
+        let l:stderr = ''
+      endif
+      let l:closure = maktaba#function#WithArgs(
```

Comment:
`call maktaba#function#Call(self.callback, [s:CurrentEnv(), l:return_data.stdout, l:stderr, 0])` is shorter. The `maktaba#function#` functions will all ensure callable anyway.